File: Week 11/881. Boats to Save People.py
class Solution:
    def numRescueBoats(self, nums: List[int], limit: int) -> int:
        nums.sort(reverse=True)
        #[50, 49, 47, 41, 33, 26, 22, 18, 13, 12, 11, 10, 10, 10, 7, 6, 6, 2, 2, 2]

        
        left  = 0
        right = len(nums)-1
        result = 0
        
        while left <= right:
            sum_value = nums[left]
            if sum_value + nums[right] <= limit:
                sum_value += nums[right]
                right -=1
            
            left += 1
            
            result += 1
        return result
            
            
            
        

File: 881-boats-to-save-people/881-boats-to-save-people.java
class Solution {
    public int numRescueBoats(int[] people, int limit) {
        //3,3,4,5
        // min_number of boot = 1
        // max_number of boot = people.length
        // 1, 2, 2,  3
        //
        Arrays.sort(people);
        
        int left = 0;
        int right = people.length - 1;
        
        int number_boot = 0;
        
        while (left <= right) {
            
            if (people[left] + people[right] > limit) {
                right -= 1;
            }
            else {
                left += 1;
                right -=1;
            }
            number_boot += 1;

        }
        
        return number_boot;
    }
}